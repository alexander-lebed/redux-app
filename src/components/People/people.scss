@import '../../../css/variables';

.search-container {
  padding-bottom: 5px;
  align-items: baseline;
  color: $secondary-color;
}

.search-input {
  font-size: 16px !important;
  border: none !important;
  box-shadow: none !important;
}

.last-seen {
  color: $secondary-color;
  font-size: 13px;
}