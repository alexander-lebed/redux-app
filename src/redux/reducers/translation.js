// @flow
import moment from 'moment';
import { setLocale } from './locale';
import enStrings from '../../lang/en';
import ruStrings from '../../lang/ru';
import type { Action, Dispatch, Locale, Translation } from '../../types';

const actions = {
    SET_STRINGS: 'SET_STRINGS'
};

const strings = (state: Translation = enStrings, action: Action) => {
    switch (action.type) {
        case actions.SET_STRINGS: {
            return {...action.payload};
        }
        default:
            return state;
    }
};

export default strings;

export function initTranslation() {
    return (dispatch: Dispatch, getState: Function) => {
        dispatch(translate(getState().locale));
    }
}

export function translate(locale: Locale) {
    return (dispatch: Dispatch) => {

        dispatch(setLocale(locale));

        moment.locale(locale);

        if (locale === 'ru') {
            dispatch({
                type: actions.SET_STRINGS,
                payload: ruStrings
            })
        } else {
            dispatch({
                type: actions.SET_STRINGS,
                payload: enStrings
            })
        }
    }
}