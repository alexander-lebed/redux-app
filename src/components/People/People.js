// @flow
import React from 'react';
import { connect, useSelector } from 'react-redux';
import queryString from 'query-string';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Table from 'react-bootstrap/Table';
import Form from 'react-bootstrap/Form';
import Modal from 'react-bootstrap/Modal';
import InputGroup from 'react-bootstrap/InputGroup';
import Button from 'react-bootstrap/Button';
import Image from 'react-bootstrap/Image';
import Jumbotron from 'react-bootstrap/Jumbotron';
import { ONLINE_STYLE, adminEmail } from '../../constants';
import { timestampToHumanDate } from '../../utils';
import { deleteUser } from '../../redux/reducers/users';
import ConfirmationModal from '../common/ConfirmationModal';
import type { User, Translation } from '../../types';

type Props = {
    history: Object,
    user: User,
    users: Array<User>,
    translation: Translation,
    deleteUser: Function
}

type State = {
    searchText: string,
    deleteUserId: string,
    clickedUser: User | null
}

export class People extends React.Component<Props, State> {

    state: State;

    constructor(props: Props) {
        super(props);
        this.state = {
            searchText: '',
            deleteUserId: '',
            clickedUser: null
        }
    }

    showUserProfile = (user: User | null) => this.setState({clickedUser: user});

    showDeleteConfirmation = (userId: string) => this.setState({deleteUserId: userId});

    hideDeleteConfirmation = () => this.setState({deleteUserId: ''});

    deleteUser = () => {
        this.props.deleteUser(this.state.deleteUserId);
        this.hideDeleteConfirmation();
    };

    render() {
        let {users, history, translation} = this.props;
        const {PEOPLE, COMMON} = translation;
        if (this.state.searchText) {
            users = users.filter(e => e.username.toLowerCase().indexOf(this.state.searchText.toLowerCase()) !== -1)
        }
        return (
            <Container fluid>
                <Row>
                    <Col md={{span: 10, offset: 1}} lg={{span: 8, offset: 2}}>
                        <Row>
                            <Col xs={12} sm={6}>
                                <InputGroup size='sm' className='search-container'>
                                    <InputGroup.Prepend>
                                        <i className='fas fa-search' />
                                    </InputGroup.Prepend>
                                    <Form.Control
                                        placeholder={PEOPLE.SEARCH_PEOPLE}
                                        aria-label='Search people'
                                        aria-describedby='search'
                                        className='search-input'
                                        value={this.state.searchText}
                                        onChange={e => this.setState({searchText: e.target.value})}
                                    />
                                </InputGroup>
                            </Col>
                        </Row>
                        {users.length === 0
                            ?  <Jumbotron className='text-center'>{COMMON.NO_RESULTS}</Jumbotron>
                            : <Table>
                                <tbody>
                                    {users.map(user => (
                                        <UserRow
                                            key={user._id}
                                            history={history}
                                            user={user}
                                            showUserProfile={u => this.showUserProfile(u)}
                                            showDeleteConfirmation={userId => this.showDeleteConfirmation(userId)}
                                        />
                                    ))}
                                </tbody>
                            </Table>
                        }
                        {this.state.clickedUser &&
                        <Modal
                            show={this.state.clickedUser !== null}
                            className='profile-modal'
                            onHide={() => this.showUserProfile(null)}
                        >
                            <Modal.Header closeButton>
                                <Modal.Title>{this.state.clickedUser.username}</Modal.Title>
                            </Modal.Header>
                            <Modal.Body style={{textAlign: 'center'}}>
                                <Image
                                    className='profile-modal-picture'
                                    src={this.state.clickedUser.pictureUrl ? this.state.clickedUser.pictureUrl : '/images/default-profile.jpg'}
                                />
                            </Modal.Body>
                        </Modal>
                        }
                        {this.state.deleteUserId &&
                        <ConfirmationModal
                            title={COMMON.DELETE_CONFIRMATION}
                            body={PEOPLE.DELETE_CONFIRMATION}
                            onConfirm={() => this.deleteUser()}
                            onCancel={() => this.hideDeleteConfirmation()}
                        />
                        }
                    </Col>
                </Row>
            </Container>
        )
    }
}

export default connect(
    state => ({
        user: state.authentication.user,
        users: state.users.users,
        translation: state.translation
    }),
    { deleteUser }
)(People);


type UserRowProps = {
    history: Object,
    user: User,
    showUserProfile: Function,
    showDeleteConfirmation: Function
}
export const UserRow = ((props: UserRowProps) => {
    const {user, showUserProfile, showDeleteConfirmation, history} = props;
    const {loggedUser, translation} = useSelector(state => ({
        loggedUser: state.authentication.user,
        translation: state.translation
    }));

    const isAdmin = loggedUser.email === adminEmail;

    const imageStyle = user.online ? ONLINE_STYLE : {};

    const goToConversationWith = (userIds: Array<string>) => {
        const query = queryString.stringify({userIds});
        history.push(`/conversation?${query}`);
    };
    return (
        <tr key={user._id} id={user._id}>
            <td style={{padding: 10}}>
                <Row>
                    <Col xs={12} sm={6}>
                        <div
                            className='profile-picture-wrapper cursor'
                            onClick={() => showUserProfile(user)}
                        >
                            <Image
                                id={`user-picture-${user._id}`}
                                roundedCircle
                                style={imageStyle}
                                className='profile-picture'
                                src={user.pictureUrl ? user.pictureUrl : '/images/default-profile.jpg'}
                            />
                        </div>
                        <div>
                            {user.username}
                            <div className='last-seen'>
                                {!user.online && `${translation.PEOPLE.LAST_SEEN} ${timestampToHumanDate(user.lastTime, false, translation)}`}
                            </div>
                        </div>
                    </Col>
                    <Col xs={12} sm={6}>
                        <div className='pull-right pt-1 pt-sm-2'>
                            {isAdmin &&
                            <>
                                <Button
                                    id={`delete-user-${user._id}`}
                                    size='sm'
                                    variant='outline-danger'
                                    onClick={() => showDeleteConfirmation(user._id)}
                                >
                                    <i className='far fa-trash-alt' style={{marginRight: 5}} /> {translation.COMMON.DELETE}
                                </Button>
                                {' '}
                            </>
                            }
                            <Button
                                id={`write-user-${user._id}`}
                                size='sm'
                                variant='outline-dark'
                                onClick={() => goToConversationWith([user._id])}
                            >
                                <i className='far fa-paper-plane' style={{marginRight: 5}} /> {translation.PEOPLE.WRITE_MESSAGE}
                            </Button>
                        </div>
                    </Col>
                </Row>
            </td>
        </tr>
    )
});
UserRow.displayName = 'UserRow';