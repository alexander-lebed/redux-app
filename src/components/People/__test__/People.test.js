// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';
import { People } from '../People';
import translation from '../../../lang/en';


describe('<People/>', () => {

    const props = {
        user: {
            _id: '111',
            username:'Current User',
            email: 'user@gmail.com',
            password: 'password',
            pictureUrl: '',
            online: true,
            lastTime: 1519294933743,
            oauth: ''
        },
        users: [
            {
                _id: '111',
                username:'Current User',
                email: 'user@mail.com',
                password: 'user-password',
                pictureUrl: '',
                online: true,
                lastTime: 1519294933743,
                oauth: ''
            },
            {
                _id: '222',
                username:'Alice',
                email: 'alice@gmail.com',
                password: 'alice-password',
                pictureUrl: '',
                online: true,
                lastTime: 1518346740388,
                oauth: ''
            },
            {
                _id: '333',
                username:'Bob',
                email: 'bob@mail.com',
                password: 'bob-password',
                pictureUrl: '',
                online: true,
                lastTime: 1518346740388,
                oauth: ''
            }
        ],
        deleteUser: jest.fn(),
        translation: translation,
        history: {}
    };

    test('should match snapshot', () => {
        const wrapper = renderer.create(
            <People {...props} />
        );
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render people', () => {
        const wrapper = mount(
            <People {...props} />
        );
        expect(wrapper.props().users.length).toEqual(3);
    });

    xit('should route to conversation', () => {
        props.history = {
            push: sinon.spy()
        };
        const wrapper = mount(
            <People {...props} />
        );

        expect(props.history.push.calledOnce).toBe(false);
        wrapper.find('#write-user-222').simulate('click');
        expect(props.history.push.withArgs('/conversation?userIds=222').calledOnce).toBe(true);
    });
});