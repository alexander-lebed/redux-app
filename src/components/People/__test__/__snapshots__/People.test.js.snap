// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<People/> should match snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "marginLeft": 0,
      "marginRight": 0,
    }
  }
>
  <div
    className="col-md-8 col-md-offset-2 col-sm-10 col-sm-offset-1 col-xs-12 col-xs-offset-0"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-5 col-xs-12"
      >
        <div
          className="form-group"
        >
          <div
            className="inner-addon left-addon"
          >
            <span
              className="glyphicon glyphicon-search"
            />
            <input
              className="form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="Search people"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <table
      className="table"
    >
      <tbody>
        <tr
          id="111"
        >
          <td>
            <div
              className="row"
            >
              <div
                className="col-sm-6 col-xs-12"
              >
                <span
                  className="glyphicon glyphicon-user"
                  style={
                    Object {
                      "color": "#00b386",
                      "marginRight": 15,
                    }
                  }
                />
                Current User
              </div>
              <div
                className="col-sm-6 col-xs-12"
                style={
                  Object {
                    "color": "grey",
                    "fontSize": 13,
                  }
                }
              >
                <button
                  className="pull-right btn btn-sm btn-default"
                  disabled={false}
                  id="write-user-111"
                  onClick={[Function]}
                  type="button"
                >
                  Write a message
                </button>
              </div>
            </div>
          </td>
        </tr>
        <tr
          id="222"
        >
          <td>
            <div
              className="row"
            >
              <div
                className="col-sm-6 col-xs-12"
              >
                <span
                  className="glyphicon glyphicon-user"
                  style={
                    Object {
                      "color": "#00b386",
                      "marginRight": 15,
                    }
                  }
                />
                Alice
              </div>
              <div
                className="col-sm-6 col-xs-12"
                style={
                  Object {
                    "color": "grey",
                    "fontSize": 13,
                  }
                }
              >
                <button
                  className="pull-right btn btn-sm btn-default"
                  disabled={false}
                  id="write-user-222"
                  onClick={[Function]}
                  type="button"
                >
                  Write a message
                </button>
              </div>
            </div>
          </td>
        </tr>
        <tr
          id="333"
        >
          <td>
            <div
              className="row"
            >
              <div
                className="col-sm-6 col-xs-12"
              >
                <span
                  className="glyphicon glyphicon-user"
                  style={
                    Object {
                      "color": "#00b386",
                      "marginRight": 15,
                    }
                  }
                />
                Bob
              </div>
              <div
                className="col-sm-6 col-xs-12"
                style={
                  Object {
                    "color": "grey",
                    "fontSize": 13,
                  }
                }
              >
                <button
                  className="pull-right btn btn-sm btn-default"
                  disabled={false}
                  id="write-user-333"
                  onClick={[Function]}
                  type="button"
                >
                  Write a message
                </button>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    
  </div>
</div>
`;
